### PR TITLE
TYP: localization.py

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

### CODE REVIEW

File: pandas/_config/localization.py [hunk 1]
```diff
@@ -3,16 +3,24 @@
...
+)
 
 from pandas._config.config import options
 
 
 @contextmanager
-def set_locale(new_locale, lc_var: int = locale.LC_ALL):
+def set_locale(
+    new_locale: str | tuple[str, str], lc_var: int = locale.LC_ALL
+) -> Generator[str | tuple[str, str], None, None]:
```
Comment (twoertwein) [inline pandas/_config/localization.py:23]: I think this is equivalent to`Iterator[str | tuple[str, str]]`

File: pandas/tests/indexes/datetimes/test_misc.py [hunk 2]
```diff
@@ -191,8 +191,9 @@ def test_datetimeindex_accessors6(self):
         assert [d.weekofyear for d in dates] == expected
 
     # GH 12806
+    # error: Unsupported operand types for + ("List[None]" and "List[str]")
     @pytest.mark.parametrize(
-        "time_locale", [None] if tm.get_locales() is None else [None] + tm.get_locales()
+        "time_locale", [None] + (tm.get_locales() or [])  # type: ignore[operator]
```
Comment (mroeschke) [inline pandas/tests/indexes/datetimes/test_misc.py:196]: @twoertwein do you think I would be able to fix this ignore somehow?

I found https://github.com/python/mypy/issues/3933 might be related?

Comment (twoertwein) [reply pandas/tests/indexes/datetimes/test_misc.py:196]: I don't think so, would need to use a cast or create the lists not inline and add explicit type annotations.

If `tm.get_locales()` can return None, mypy might accept it if you move the `tm.get_locales()` call before `None`.

Comment (mroeschke) [reply pandas/tests/indexes/datetimes/test_misc.py:196]: Reversing the order yields `error: List item 0 has incompatible type "None"; expected "str"  [list-item]`

I think I'll just keep this ignore for now. Adding the casts may make these tests a little more verbose than necessary.

**Pull request merged**